verify.c

```diff
@@ -0,0 +1,563 @@
+		stream = &verifier->output_stream;
+	} else {
+		assert(fd == verifier->error_stream.fd);
+		stream = &verifier->error_stream;
+	}
+
+	assert(stream->fd != -1);
+
+	do {
+		/* (re)fill buffer */
+		errno = 0;
+		end = NULL;
+		cnt = read(stream->fd,
+		           stream->buf + stream->cnt,
+		           LOGBUFSIZE - stream->cnt);
+		if(cnt > 0) {
+			stream->buf[stream->cnt + (size_t)cnt] = '\0';
+			end = strchr(stream->buf + stream->cnt, '\n');
+			stream->cnt += (size_t)cnt;
+			if(end == NULL && stream->cnt == LOGBUFSIZE) {
+				end = stream->buf + stream->cnt;
+			}
+		} else if((cnt ==  0) ||
+		          (cnt == -1 && !(errno == EAGAIN || errno == EINTR)))
+		{
+			/* flush what is available */
+			end = stream->cnt > 0 ? stream->buf + stream->cnt : NULL;
+		}
+		while(end != NULL) {
+			ptr = stream->buf;
+			len = end - ptr;
+			while(len > LOGLINELEN) {
+				fmt = stream->cut ? ".. %.*s .." : "%.*s ..";
+				stream->cut = 1; /* split line */
+				log_msg(stream->priority, fmt, LOGLINELEN, ptr);
+				ptr += LOGLINELEN;
+				len -= LOGLINELEN;
+			}
+			assert(len <= LOGLINELEN);
+			fmt = stream->cut ? ".. %.*s" : "%.*s";
+			if(*end == '\n' || cnt <= 0) {
+				stream->cut = 0;
+				/* flush buffer unless more data is expected */
+				*end = '\0';
+				log_msg(stream->priority, fmt, len, ptr);
+				ptr += len;
+			}
+			if(cnt <= 0) {
+				break;
+			}
+			/* move data to head of buffer */
+			len = (ptr - stream->buf);
+			memmove(stream->buf, ptr + 1, (stream->cnt - len) - 1);
+			stream->cnt -= len + 1;
+			stream->buf[stream->cnt] = '\0';
+			end = stream->cnt > 0 ? strchr(stream->buf, '\n') : NULL;
+		}
+	} while(cnt > 0);
+
+	if(cnt <= 0 && !(errno == EAGAIN || errno == EINTR)) {
+		event_del(&stream->event);
+		close(stream->fd);
+		stream->fd = -1;
+	}
+}
+
+static void kill_verifier(struct verifier *verifier)
+{
+	assert(verifier != NULL);
+	assert(verifier->zone != NULL);
+
+	if(kill(verifier->pid, SIGTERM) == -1) {
+		log_msg(LOG_ERR, "verify: cannot kill verifier for "
+		                 "zone %s (pid %d): %s",
+		                 verifier->zone->opts->name,
+		                 verifier->pid,
+		                 strerror(errno));
+	}
+}
+
+static void close_verifier(struct verifier *verifier)
+{
+	/* unregister events and close streams (in that order) */
+	if(verifier->timeout.tv_sec > 0) {
+		event_del(&verifier->timeout_event);
+		verifier->timeout.tv_sec = 0;
+		verifier->timeout.tv_usec = 0;
+	}
+
+	if(verifier->zone_feed.fh != NULL) {
+		event_del(&verifier->zone_feed.event);
+		fclose(verifier->zone_feed.fh);
+		verifier->zone_feed.fh = NULL;
+		region_destroy(verifier->zone_feed.region);
+	}
+
+	if (verifier->output_stream.fd != -1) {
+		verify_handle_stream(
+			verifier->output_stream.fd, EV_READ, verifier);
+		verifier->output_stream.fd = -1;
+	}
+
+	if (verifier->error_stream.fd != -1) {
+		verify_handle_stream(
+			verifier->error_stream.fd, EV_READ, verifier);
+		verifier->output_stream.fd = -1;
```

Comment:
s/output_stream/error_stream/
But it does not matter because the routine should have set it to -1, but it is still fine to set it here in case it hasn't.  It would then also be a good idea to close that fd here, if you think it is not always -1.

Reply:
done